### PR TITLE
Use a struct instead of a tuple for iteration

### PR DESCRIPTION
**Pull request by marshallpierce**
This seems less error prone to work with, allows us to centralize the docs on what each component means, and will also allow us to unobtrusively add more fields. Performance shouldn't be affected since tuples are just anonymous structs, AFAIK.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -800,23 +801,27 @@ impl<T: Counter> Histogram<T> {
     ///
     /// println!("{:?}", hist.iter_percentiles(1).collect::<Vec<_>>());
     ///
-    /// assert_eq!(perc.next(), Some((hist.value_at_percentile(0.01), 0.01, 1, 1)));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: hist.value_at_percentile(0.01),
+    ///     percentile: 0.01, count_at_value: 1, count_since_last_iteration: 1 }));
```
Comment (jonhoo) [inline src/lib.rs]: The wrapping here isn't great. Can you make it more struct-y?

File: src/lib.rs [hunk 2]
```diff
@@ -800,23 +801,27 @@ impl<T: Counter> Histogram<T> {
...
+    /// assert_eq!(perc.next(), Some(IterationValue { value: hist.value_at_percentile(96.88),
+    ///     percentile: 96.88, count_at_value: 1, count_since_last_iteration: 313 }));
     /// // etc...
     /// ```
-    pub fn iter_percentiles<'a>
-        (&'a self,
-         percentileTicksPerHalfDistance: isize)
-         -> iterators::HistogramIterator<'a, T, iterators::percentile::Iter<'a, T>> {
+    pub fn iter_percentiles<'a>(&'a self, percentileTicksPerHalfDistance: isize)
+            -> HistogramIterator<'a, T, iterators::percentile::Iter<'a, T>> {
```
Comment (jonhoo) [inline src/lib.rs]: This is just a `rustfmt` change, right?

File: src/lib.rs (hunk 1 above)
Comment (jonhoo) [inline src/lib.rs]: Alternatively, give `IterationValue` a `new` method or an `impl From<(...)>`

File: src/lib.rs [hunk 3]
```diff
@@ -825,96 +830,102 @@ impl<T: Counter> Histogram<T> {
...
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 499,
+    ///     percentile: hist.percentile_below(499), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 599,
+    ///     percentile: hist.percentile_below(599), count_at_value: 0, count_since_last_iteration: 1 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 699,
+    ///     percentile: hist.percentile_below(699), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 799,
+    ///     percentile: hist.percentile_below(799), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 899,
+    ///     percentile: hist.percentile_below(899), count_at_value: 0, count_since_last_iteration: 2 }));
```
Comment (jonhoo) [inline src/lib.rs]: Formatting here too.

File: src/lib.rs [hunk 4]
```diff
@@ -825,96 +830,102 @@ impl<T: Counter> Histogram<T> {
...
-    /// assert_eq!(perc.next(), Some((99, hist.percentile_below(99), 0, 0)));
-    /// assert_eq!(perc.next(), Some((999, hist.percentile_below(999), 0, 4)));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 0,
+    ///     percentile: hist.percentile_below(0), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 9,
+    ///     percentile: hist.percentile_below(9), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 99,
+    ///     percentile: hist.percentile_below(99), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 999,
+    ///     percentile: hist.percentile_below(999), count_at_value: 0, count_since_last_iteration: 4 }));
```
Comment (jonhoo) [inline src/lib.rs]: Formatting.

File: src/lib.rs [hunk 5]
```diff
@@ -825,96 +830,102 @@ impl<T: Counter> Histogram<T> {
...
-    /// assert_eq!(perc.next(), Some((800, hist.percentile_below(800), 1, 1)));
-    /// assert_eq!(perc.next(), Some((850, hist.percentile_below(850), 1, 1)));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 100,
+    ///     percentile: hist.percentile_below(100), count_at_value: 1, count_since_last_iteration: 1 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 500,
+    ///     percentile: hist.percentile_below(500), count_at_value: 1, count_since_last_iteration: 1 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 800,
+    ///     percentile: hist.percentile_below(800), count_at_value: 1, count_since_last_iteration: 1 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 850,
+    ///     percentile: hist.percentile_below(850), count_at_value: 1, count_since_last_iteration: 1 }));
```
Comment (jonhoo) [inline src/lib.rs]: Formatting.

File: src/lib.rs [hunk 6]
```diff
@@ -923,32 +934,41 @@ impl<T: Counter> Histogram<T> {
...
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 6,
+    ///     percentile: hist.percentile_below(6), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 7,
+    ///     percentile: hist.percentile_below(7), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 8,
+    ///     percentile: hist.percentile_below(8), count_at_value: 1, count_since_last_iteration: 1 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 9,
+    ///     percentile: hist.percentile_below(9), count_at_value: 0, count_since_last_iteration: 0 }));
+    /// assert_eq!(perc.next(), Some(IterationValue { value: 10,
+    ///     percentile: 100.0, count_at_value: 0, count_since_last_iteration: 0 }));
```
Comment (jonhoo) [inline src/lib.rs]: Formatting.

File: src/lib.rs (hunk 2 above)
Comment (jonhoo) [inline src/lib.rs]: Hmm, no, it changed from `iterators::HistogramIterator` to just `HistogramIterator`. Is this a necessary change? I mean, I prefer it this new way with a `use iterators::HistogramIterator` at the top, but that should be a separate PR.

File: src/iterators/mod.rs [hunk 7]
```diff
@@ -47,6 +47,19 @@ pub struct HistogramIterator<'a, T: 'a + Counter, P: PickyIterator<T>> {
...
+    /// the lowest value stored in the current histogram bin
+    pub value: u64,
+    /// percent of recorded values that are equivalent to or below `value`
+    pub percentile: f64,
+    /// recorded count for values equivalent to `value`
+    pub count_at_value: T,
+    /// number of values traversed since the last iteration step
+    pub count_since_last_iteration: u64
+}
+
```
Comment (jonhoo) [inline src/iterators/mod.rs]: I think I'd prefer if there weren't `pub`, and instead there were `&self` accessors. We never expect an `IterationValue` to have its internals modified. We'll also need an `impl` block anyway for `::new`.

**Comment by jonhoo** [comment]
I like this change, though see the inline comments. Also needs to be rebased after I merged #4.

File: src/lib.rs (hunk 1 above)
Comment (marshallpierce) [inline src/lib.rs]: I'm happy to reformat; I only went with compact-but-ugly to save relatively boring lines of source. Would you prefer more like this:

```
    /// assert_eq!(perc.next(), Some(IterationValue { 
    ///     value: hist.value_at_percentile(0.01),
    ///     percentile: 0.01, 
    ///     count_at_value: 1,
    ///     count_since_last_iteration: 1
    /// }));
￼ ```

Given the documentation nature of these asserts, I think it's nice for the user to see explicitly what data is available at each point in the iteration, but if we move to non-`pub` members then we'll have `new` here anyway.

File: src/lib.rs (hunk 2 above)
Comment (marshallpierce) [inline src/lib.rs]: It was visually jarring so I just fixed it by hand (here and the few other function param re-wrappings). I added the `use` to shorten the line with `HistogramIterator` on it, but I can pull that out to a separate PR.

File: src/iterators/mod.rs (hunk 7 above)
Comment (marshallpierce) [inline src/iterators/mod.rs]: Yeah, I was on the fence about that. Since you're in favor I'll go ahead and do it. I wish it was possible to have the equivalent of Java's `public final`. :(

Comment (marshallpierce) [inline src/iterators/mod.rs]: One possibility would be to use something like https://play.rust-lang.org/?code=%0Ause%20immutable%3A%3AImmutable%3B%0A%0Astruct%20Foo%20%7B%20bar%3A%20u64%2C%20baz%3A%20u64%20%7D%0A%0Aimpl%20Foo%20%7B%0A%20%20%20%20fn%20sum(%26self)%20-%3E%20u64%20%7B%0A%20%20%20%20%20%20%20%20self.bar%20%2B%20self.baz%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20fn%20reset(%26mut%20self)%20%7B%0A%20%20%20%20%20%20%20%20self.bar%20%3D%200%3B%0A%20%20%20%20%20%20%20%20self.baz%20%3D%200%3B%0A%20%20%20%20%7D%0A%7D%0A%0Afn%20get_foo()%20-%3E%20Immutable%3CFoo%3E%20%7B%0A%20%20%20%20Immutable%3A%3Anew(Foo%20%7B%20bar%3A%201%2C%20baz%3A%201%20%7D)%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20x%20%3D%20get_foo()%3B%0A%20%20%20%20%0A%20%20%20%20println!(%22%7B%7D%20%2B%20%7B%7D%20%3D%20%7B%7D%22%2C%20x.bar%2C%20x.baz%2C%20x.sum())%3B%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20x.reset()%3B%20%2F%2F%20error%3A%20cannot%20borrow%20immutable%20borrowed%20content%20as%20mutable%0A%20%20%20%20%2F%2F%20x.bar%20%3D%205%3B%20%2F%2F%20error%3A%20cannot%20assign%20to%20immutable%20field%0A%7D%0A%0Amod%20immutable%20%7B%0A%20%20%20%20use%20std%3A%3Aops%3A%3ADeref%3B%0A%20%20%20%20%0A%20%20%20%20pub%20struct%20Immutable%3CT%3E(T)%3B%0A%20%20%20%20%0A%20%20%20%20impl%3CT%3E%20Immutable%3CT%3E%20%7B%0A%20%20%20%20%20%20%20%20pub%20fn%20new(t%3A%20T)%20-%3E%20Self%20%7B%20Immutable(t)%20%7D%0A%20%20%20%20%20%20%20%20pub%20fn%20get(%26self)%20-%3E%20%26T%20%7B%20%26self.0%20%7D%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20impl%3CT%3E%20Deref%20for%20Immutable%3CT%3E%20%7B%0A%20%20%20%20%20%20%20%20type%20Target%20%3D%20T%3B%0A%20%20%20%20%20%20%20%20fn%20deref(%26self)%20-%3E%20%26T%20%7B%20%26self.0%20%7D%0A%20%20%20%20%7D%0A%7D%0A&version=stable&backtrace=0.

Comment (marshallpierce) [inline src/iterators/mod.rs]: After considering, I think the immutability-via-wrapper-struct is a bit heavy-handed because it would be a confusing thing for users to see and would also require them to have `Immutable<Foo>` in the type if they were to store these values in a collection, for instance. Ugly.

File: src/lib.rs (hunk 1 above)
Comment (jonhoo) [inline src/lib.rs]: Yes, that's the style I was proposing, though I think `new` would be even better. I recognize that the full `struct` constructor conveys more information for documentation purposes, though I think the version using `new` will be more readable, and possibly more understandable.

File: src/iterators/mod.rs (hunk 7 above)
Comment (jonhoo) [inline src/iterators/mod.rs]: Yeah, `Immutable` is a neat pattern, but it seems like overkill here.

**Comment by marshallpierce** [comment]
OK I think that addresses all your feedback.

File: src/iterators/mod.rs [hunk 8]
```diff
@@ -50,14 +50,43 @@ pub struct HistogramIterator<'a, T: 'a + Counter, P: PickyIterator<T>> {
...
+    value: u64,
+    percentile: f64,
+    count_at_value: T,
+    count_since_last_iteration: u64
+}
+
+impl<T: Counter> IterationValue<T> {
+    /// Create a new IterationValue.
+    pub fn new(value: u64, percentile: f64, count_at_value: T, count_since_last_iteration: u64)
+            -> IterationValue<T> {
```
Comment (jonhoo) [inline src/iterators/mod.rs]: I believe you can use `-> Self` here

**Comment by jonhoo** [comment]
LGTM, thanks!
